src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -192,101 +122,254 @@ LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
->
-double FFN<
-LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
->::Evaluate(const arma::mat& /* unused */,
-            const size_t i,
-            const bool deterministic)
+template<typename OutputLayerType, typename InitializationRuleType>
+double FFN<OutputLayerType, InitializationRuleType>::Evaluate(
+    const arma::mat& /* parameters */, const size_t i, const bool deterministic)
 {
-  this->deterministic = deterministic;
+  if (parameter.is_empty())
+  {
+    ResetParameters();
+  }
 
-  ResetParameter(network);
+  if (deterministic != this->deterministic)
+  {
+    this->deterministic = deterministic;
+    ResetDeterministic();
+  }
 
-  Forward(arma::mat(predictors.colptr(i), predictors.n_rows, 1, false, true),
-      network);
+  currentInput = std::move(arma::mat(predictors.colptr(i),
+      predictors.n_rows, 1, false, true));
 
-  return OutputError(arma::mat(responses.colptr(i), responses.n_rows, 1, false,
-      true), error, network);
+  Forward(std::move(currentInput));
+
+  currentTarget = arma::mat(responses.colptr(i), responses.n_rows,
+      1, false, true);
+
+  double res = outputLayer.Forward(std::move(boost::apply_visitor(
+      outputParameterVisitor, network.back())), std::move(currentTarget));
+
+  return res;
 }
 
-template<typename LayerTypes,
-         typename OutputLayerType,
-         typename InitializationRuleType,
-         typename PerformanceFunction
->
-void FFN<
-LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
->::Gradient(const arma::mat& /* unused */,
-            const size_t i,
-            arma::mat& gradient)
+template<typename OutputLayerType, typename InitializationRuleType>
+void FFN<OutputLayerType, InitializationRuleType>::Gradient(
+    const arma::mat& parameters, const size_t i, arma::mat& gradient)
 {
   if (gradient.is_empty())
   {
+    if (parameter.is_empty())
+    {
+      ResetParameters();
+    }
+
     gradient = arma::zeros<arma::mat>(parameter.n_rows, parameter.n_cols);
   }
+  else
+  {
+    gradient.zeros();
+  }
+
+  Evaluate(parameters, i, false);
+
+  outputLayer.Backward(std::move(boost::apply_visitor(outputParameterVisitor,
+      network.back())), std::move(currentTarget), std::move(error));
 
+  Backward();
+  ResetGradients(gradient);
+  Gradient();
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+void FFN<OutputLayerType, InitializationRuleType>::ResetParameters()
+{
+  size_t weights = 0;
+  for (size_t i = 0; i < network.size(); ++i)
+  {
+    weights += boost::apply_visitor(weightSizeVisitor, network[i]);
+  }
 
-  Evaluate(parameter, i, false);
+  parameter.set_size(weights, 1);
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
 
-  NetworkGradients(gradient, network);
+  size_t offset = 0;
+  for (size_t i = 0; i < network.size(); ++i)
+  {
+    offset += boost::apply_visitor(WeightSetVisitor(std::move(parameter),
+        offset), network[i]);
```

Comment:
Right, the drawback is, that it's not that easy to change the weight vector size during training, but I couldn't think of any situation where that's necessary.